### PR TITLE
#167729787 User Login Endpoint

### PR DESCRIPTION
**Pull request by chibytez**
#### What does this PR do?
Adding the user login endpoint
#### Description of Task to be completed?
added the login controller
added the login router
validated the email and password
wrote tested for the login endpoint
#### How should this be manually tested?
can be tested with postman
#### Any background context you want to provide?
none

#### What are the relevant pivotal tracker stories?
  #167729787
#### Screenshots (if appropriate)
Capture 1
Capture 2

#### Questions:

### CODE REVIEW

**Review by okabamac — APPROVED** [review]
Great work, it's OK. I've got a utility function in my PR to help reduce the res.status.json settings everywhere. Well done.

**Review by sebaztyn — APPROVED** [review]
Good work.

File: package.json [hunk 1]
```diff
@@ -14,11 +14,12 @@
     "migration": "sequelize db:migrate",
     "seed": "sequelize db:seed:all",
     "reverse": "sequelize db:migrate:undo:all",
+    "refreshdb": "sequelize db:migrate:undo:all && sequelize db:migrate",
     "coveralls-coverage": "coveralls < lcov.info",
     "coverage": "nyc npm test && npm run generate-lcov && npm run coveralls-coverage && npm run codeclimate-coverage",
     "codeclimate-coverage": "codeclimate-test-reporter < lcov.info",
     "generate-lcov": "nyc report --reporter=text-lcov > lcov.info",
-    "test": "cross-env NODE_ENV=test mocha --require @babel/register src/test/**.test.js --timeout 20000 --exit"
+    "test": "cross-env NODE_ENV=test npm run heroku-postbuild && nyc _mocha --require @babel/register  src/test/**.test.js --timeout 20000 --exit"
```
Comment (Stylll) [inline package.json]: Is there a reason we are running `heroku-postbuild` when the test script runs?

File: src/controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import bcrypt from 'bcryptjs';
+import jwt from 'jsonwebtoken';
+import models from '../models';
+
+const { User } = models;
+
+/**
```
[the remark this answers is not in the document]
Comment (Stylll) [reply src/controllers/userController.js:6]: Please fix this.

File: src/controllers/userController.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        });
+      }
+      const payload = {
+        id: user.id,
+        email: user.email
+      };
+      jwt.sign(
+        { id: user.id, email: user.email },
+        process.env.SECRET, (err, token) => res.status(201).json({
+          status: 'success', message: 'User successfully logged in', payload, token
```
Comment (Stylll) [inline src/controllers/userController.js]: The status code here shouldn't be `201` as we are not creating the user. Also, we need to handle cases where the token is empty or an error occurred while signing.

File: src/test/login.test.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      .end((err, res) => {
+        expect(res.status).to.equal(201);
+        expect(res.body.payload).to.be.an('object');
+        expect(res.body).to.have.property('status');
+        expect(res.body.message).to.equal('User successfully logged in');
+        done();
+      });
+  });
+
+  it('should return an invalid login', (done) => {
```
Comment (Stylll) [inline src/test/login.test.js]: Can you make this more descriptive?. `should return an invalid login when ....`

File: src/validation/userValidation.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { check } from 'express-validator';
+
+const loginValidation = [
+  check('email').isEmail().trim().withMessage('input a valid email address'),
```
Comment (Stylll) [inline src/validation/userValidation.js]: Please update the first letter of each message to capital letter

**Review by Stylll — CHANGES_REQUESTED** [review]
Hey @chibytez, I dropped a few comments. Kindly check them out.

File: package.json (hunk 1 above)
Comment (chibytez) [reply package.json]: yeah, so it can undo migration, run migration and seed the test 
database

File: src/test/login.test.js (hunk 4 above)
Comment (chibytez) [reply src/test/login.test.js]: ok thanks

File: src/controllers/userController.js (hunk 3 above)
Comment (chibytez) [reply src/controllers/userController.js]: ok thank

File: src/validation/userValidation.js (hunk 5 above)
Comment (chibytez) [reply src/validation/userValidation.js]: ok thanks

File: src/validation/errorMessage.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { validationResult } from 'express-validator/check';
+
+const Errs = {
```
Comment (Stylll) [inline src/validation/errorMessage.js]: This variable name isn't descriptive enough. Can we use `Errors` or something else that'll easily tell what the object is?

File: package.json (hunk 1 above)
Comment (Stylll) [reply package.json]: Okay. That script is strictly for `heroku-postbuild`, we shouldn't use it here so we don't call it unnecessary if it is later updated with something specifically for Heroku. You can call the scripts to undo migration, run migration and seed test here or create a new script combining them together.

**Review by Stylll — CHANGES_REQUESTED** [review]
@chibytez Just a few changes. Please check them out.

File: src/validation/errorMessage.js (hunk 6 above)
Comment (chibytez) [reply src/validation/errorMessage.js]: ok thank, have implemented it

File: package.json (hunk 1 above)
Comment (chibytez) [reply package.json]: ok have implemented it

**Review by mthamayor — APPROVED** [review]
Nice work

**Review by okabamac — APPROVED** [review]
Nice work in implementing all the changes. Looks good

File: src/routes/api/index.js [hunk 7]
```diff
@@ -1,31 +1,18 @@
...
-  }
-  passport.authenticate('local', { session: false }, (
-    err,
-    user,
-    info
-  ) => {
-    if (err) {
-      return next(err);
-    }
+usersRouter.use('/auth', authRouter);
```
Comment (Stylll) [inline src/routes/api/index.js]: We can prefix this route with the version. `v1`

**Review by Stylll — CHANGES_REQUESTED** [review]
I noticed the error message is in this format when no email or password is passed.
```
{
    "status": 422,
    "errors": [
        "Input a valid email address",
        "Input a valid email address",
        "nonandela email",
        "Input password",
        "Input longer password"
    ]
}
```

Can we have it in a way where it lets the user know what fields are having the error

```
{
    "status": 422,
    "errors": {
       "email": "Input a valid email address",
       "password": "password is required"
    }
}
```

OR

```
{
    "status": 422,
    "errors": {
       "email": ["Input a valid email address", "email should be an andela email"],
       "password": ["password is required"]
    }
}
```

File: src/routes/api/index.js (hunk 7 above)
Comment (chibytez) [reply src/routes/api/index.js]: thanks a lot for reviewing my work, have implemented the required changes

**Review by Stylll — CHANGES_REQUESTED** [review]
Hey @chibytez, I noticed the error message is in this format 
```
{
    "status": 422,
    "errors": [
        {
            "email": "Input email"
        },
        {
            "email": "Input a valid email address"
        },
        {
            "email": "Andela email required"
        },
        {
            "password": "Input password"
        },
        {
            "password": "Input longer password"
        }
    ]
}
```

This is not the format I specified. Kindly update it.

File: src/validation/userValidation.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    .exists({ checkFalsy: true })
+    .withMessage(`email: ${loginErrors.undefinedEmail}`)
+    .isEmail()
+    .withMessage(`email: ${loginErrors.invalidEmail}`)
+    .matches(/@andela.com$/)
+    .withMessage(`email: ${loginErrors.nonAndelanEmail}`),
+  check('password')
+    .exists({ checkFalsy: true })
+    .withMessage(`password: ${loginErrors.undefinedPassword}`)
+    .isLength({ min: 8 })
```
Comment (Stylll) [inline src/validation/userValidation.js]: I don't think we need to validate if the user's password is up to 8 characters when logging in. Please remove this.

**Review by Stylll — CHANGES_REQUESTED** [review]
Hey @chibytez, please pay attention to the details of the error message format I specified. What you have done still doesn't match that format. Please reach out to @mthamayor since he has implemented something similar.
I left a comment on the validations also, please check it out.
Kindly change the label of this PR to `Peer Review` and request for review from your peers after implementing the requested changes.

File: src/validation/userValidation.js (hunk 8 above)
Comment (chibytez) [reply src/validation/userValidation.js]: ok will remove it now

Comment (chibytez) [reply src/validation/userValidation.js]: thanks for the feedback, will follow the instructions

**Review by mthamayor — CHANGES_REQUESTED** [review]
The tests are failing locally. Maybe you'll need to append 'v1' to your test URLs, changing `/api/auth/login` to `/api/v1/auth/login`

**Review by mthamayor — CHANGES_REQUESTED** [review]
One of the tests is failing. Please check them.

**Review by okabamac — COMMENTED** [review]
I can only get one test to pass locally. Tell me if there's something I'm doing wrong. It's sort of throwing a 400 where we are expecting a 404 error and if the spec is followed properly that should be a validation error.

**Review by davidfrank96 — APPROVED** [review]
Looks Good

**Review by okafor-chidimma — CHANGES_REQUESTED** [review]
Good work @chibytez ...however these are the changes that would aid the test
1.  Replicate every environmental variable in the .env.example. for e.g i did not know i had to declare an environmental variable called SECRET before testing
2. ensure that the arrow-parens property is turned off in your eslint.json
3. ensure that the db tablename is in Plural to avoid errors such as "could not find relation: User"

When I effected these changes locally, the tests were passing

**Review by Stylll — APPROVED** [review]

**Pull request merged**